qlf_fasm/qlf_fasm.py

```diff
@@ -0,0 +1,513 @@
+#!/usr/bin/env python3
+"""
+QuickLogic qlf-series FPGA FASM to bitstream and bitstream to FASM conversion
+utility.
+"""
+import argparse
+import os
+import re
+import logging
+import pkg_resources
+
+import fasm
+
+from .database import Bit, Database
+from .bitstream import TextBitstream
+
+# =============================================================================
+
+DATABASES = {
+    "qlf_k4n8": pkg_resources.resource_filename("qlf_fasm", "database/qlf_k4n8")
+}
+
+# =============================================================================
+
+
+class QlfFasmAssembler():
+    """
+    FASM assembler for QuickLogic QLF devices.
+    """
+
+    LOC_RE = re.compile(r"(?P<name>.+)_(?P<x>[0-9]+)__(?P<y>[0-9]+)_$")
+
+    class LookupError(Exception):
+        """
+        FASM database lookup error exception
+        """
+        pass
+
+    class FeatureConflict(Exception):
+        """
+        FASM feature conflict exception
+        """
+        pass
+
+    def __init__(self, database):
+        self.bitstream = bytearray(database.bitstream_size)
+        self.database = database
+
+        self.features_by_bits = {}
+
+    def process_fasm_line(self, line):
+        """
+        Assembles and updates a part of the bistream described by the given
+        single FASM line object.
+        """
+
+        set_feature = line.set_feature
+        if not set_feature:
+            return
+
+        # Ignore features that are not set
+        if set_feature.value == 0:
+            return
+
+        # Split the feature name into parts, check the first part
+        parts = set_feature.feature.split(".")
+        if len(parts) < 3 or parts[0] != "fpga_top":
```

Comment:
`fpga_top` is used in multiple locations. Might be worth using a top-file variable to define this string?

Reply:
So this is a constant string specific to bit names in the OpenFPGA project. Instead of adding an option to change it I created a global variable which holds the na,e.